### ISSUE TITLE
scala-native Scala 3 support

### ISSUE THREAD

**Issue by xuwei-k** [scalatest/scalatest#2097]
🙏 

- https://twitter.com/scala_native/status/1484889552833687552
- https://scala-native.readthedocs.io/en/v0.4.3/changelog/0.4.3.html

**Comment by cheeseng** [comment]
@xuwei-k  Fyi we are unable to release Scala 3 build for scala-native yet, but I have bumped up scala-native used to build artifacts for 3.2.11 and staged them.  We'll work on add on build for Scala 3 scala-native for 3.2.11.

**Comment by cheeseng** [comment]
@xuwei-k Unfortunately this is now blocked by scala-xml 2.0.1 not being published for Scala 3 native: 

https://mvnrepository.com/artifact/org.scala-lang.modules/scala-xml

@SethTisue Any chance the scala-xml build for Scala 3 native can be published?

Thanks.

**Comment by cheeseng** [comment]
@xuwei-k @SethTisue Seems like this needs to be completed first: 

https://github.com/scala/scala-xml/pull/584

**Comment by SethTisue** [comment]
I can't really afford to learn about Scala Native details myself, but I can certainly roll a scala-xml release once Yoshida-san and/or other volunteers finish that PR.

**Comment by sideeffffect** [comment]
There's a PR now for scala-xml, just waiting to be merged https://github.com/scala/scala-xml/pull/592

**Comment by catap** [comment]
@cheeseng `scala-xml-2.1.0` was released with support of scala3 at scala-native.

May I ask you to move forward here? Thanks!

**Comment by catap** [comment]
Here it is: https://github.com/scalatest/scalatest/pull/2114

**Comment by cheeseng** [comment]
@catap Sure, I'll work on it, thanks for informing!

**Comment by catap** [comment]
@cheeseng may I ask you which commit / branch have you used to test it?

**Comment by cheeseng** [comment]
@catap Sure, I am staging 3.2.12 from this branch of my fork, in this PR: 

https://github.com/scalatest/scalatest/pull/2121

and the exact branch URL is: 

https://github.com/cheeseng/scalatest/tree/feature-3.2.12-rel

It is late here so I'll have to call it a day, hopefully we can continue to publish 3.2.12 that includes a working scala-native build for scala 3 tomorrow!  

Thanks!  :)

**Comment by catap** [comment]
@cheeseng the first idea is https://github.com/cheeseng/scalatest/pull/4 can I ask you to try it?

**Comment by cheeseng** [comment]
@catap Sure I'll try it a try today, thanks for the PR!

**Comment by cheeseng** [comment]
@catap Hmm, unfortunately it still failed with the same error: 

```
[error] Found 8 missing definitions while linking
[error] Not found Top(org.scalatest.SuitesL20org.scalatest.SuitesL30scala.collection.immutable.Seq$scalanative$ReflectivelyInstantiate$)
[error] Not found Member(Top(org.scalatest.SuitesL20org.scalatest.SuitesL30scala.collection.immutable.Seq$scalanative$ReflectivelyInstantiate$),RE)
[error] 	at /home/cheeseng/git/scalatest/dotty/core.native/target/scala-3/src_managed/main/org/scalatest/Suites.scala:57
[error] Not found Top(org.scalatest.funspec.AnyFunSpecL32org.scalatest.funspec.AnyFunSpec$scalanative$ReflectivelyInstantiate$)
[error] Not found Member(Top(org.scalatest.funspec.AnyFunSpecL32org.scalatest.funspec.AnyFunSpec$scalanative$ReflectivelyInstantiate$),RE)
[error] 	at /home/cheeseng/git/scalatest/dotty/funspec.native/target/scala-3/src_managed/main/org/scalatest/funspec/AnyFunSpec.scala:1631
[error] Not found Top(org.scalatest.funsuite.AnyFunSuiteL34org.scalatest.funsuite.AnyFunSuite$scalanative$ReflectivelyInstantiate$)
[error] Not found Member(Top(org.scalatest.funsuite.AnyFunSuiteL34org.scalatest.funsuite.AnyFunSuite$scalanative$ReflectivelyInstantiate$),RE)
[error] 	at /home/cheeseng/git/scalatest/dotty/funsuite.native/target/scala-3/src_managed/main/org/scalatest/funsuite/AnyFunSuite.scala:1563
[error] Not found Top(org.scalatest.tools.FrameworkL29org.scalatest.tools.Framework$scalanative$ReflectivelyInstantiate$)
[error] Not found Member(Top(org.scalatest.tools.FrameworkL29org.scalatest.tools.Framework$scalanative$ReflectivelyInstantiate$),RE)
[error] 	at /home/cheeseng/git/scalatest/dotty/core.native/target/scala-3/src_managed/main/org/scalatest/tools/Framework.scala:24
[error] Undefined definitions found in reachability phase
[error] (Test / nativeLink) Undefined definitions found in reachability phase
```

Build for Scala 2.13 and 2.12 still works though.

**Comment by catap** [comment]
@cheeseng can you share exactly command which have you used to have this error? `sbt scalatestTestDottyNative/test` works fine on my side, but I may have some caching / etc somewhere, and if it the command I'll clone the fresh repo to try it one more time.

**Comment by cheeseng** [comment]
@catap Yes sbt scalatestTestDottyNative/test should work fine, I hit the error only when I tried to use the artifacts I staged in Sonatype from scalatestplus-scalacheck in this branch: 

https://github.com/scalatest/scalatestplus-scalacheck/tree/feature-3.2.12.0-rel

to get the error you can run: 

```
sbt ++3.1.2 scalatestPlusScalaCheckNative/test
```

To test that you'll need to add the staged repo in your ~/.sbt/ , can I email the repo urls to [kirill@korins.ky](mailto:kirill@korins.ky) ?

**Comment by catap** [comment]
Yep, this is my email, feel free to use it

**Comment by cheeseng** [comment]
@catap I just sent you the repo url, can you check your inbox?

**Comment by catap** [comment]
@cheeseng an attempt to cal `sbt ++3.1.2 scalatestPlusScalaCheckNative/test ` fails with error:
```
java.lang.RuntimeException: Switch failed: no subprojects list "3.1.2" (or compatible version) in crossScalaVersions setting.
If you want to force it regardless, call ++ 3.1.2!
	at scala.sys.package$.error(package.scala:30)
	at sbt.Cross$.switchScalaVersion(Cross.scala:334)
	at sbt.Cross$.switchCommandImpl(Cross.scala:244)
	at sbt.Cross$.$anonfun$switchVersion$2(Cross.scala:241)
	at sbt.Command$.$anonfun$applyEffect$4(Command.scala:150)
	at sbt.Command$.$anonfun$applyEffect$2(Command.scala:145)
	at sbt.Command$.process(Command.scala:189)
	at sbt.MainLoop$.$anonfun$processCommand$5(MainLoop.scala:243)
	at scala.Option.getOrElse(Option.scala:189)
	at sbt.MainLoop$.process$1(MainLoop.scala:243)
	at sbt.MainLoop$.processCommand(MainLoop.scala:274)
	at sbt.MainLoop$.$anonfun$next$5(MainLoop.scala:163)
	at sbt.State$StateOpsImpl$.runCmd$1(State.scala:289)
	at sbt.State$StateOpsImpl$.process$extension(State.scala:325)
	at sbt.MainLoop$.$anonfun$next$4(MainLoop.scala:163)
	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
	at sbt.MainLoop$.next(MainLoop.scala:163)
	at sbt.MainLoop$.run(MainLoop.scala:144)
	at sbt.MainLoop$.$anonfun$runWithNewLog$1(MainLoop.scala:119)
	at sbt.io.Using.apply(Using.scala:27)
	at sbt.MainLoop$.runWithNewLog(MainLoop.scala:112)
	at sbt.MainLoop$.runAndClearLast(MainLoop.scala:66)
	at sbt.MainLoop$.runLoggedLoop(MainLoop.scala:51)
	at sbt.MainLoop$.runLogged(MainLoop.scala:42)
	at sbt.StandardMain$.runManaged(Main.scala:193)
	at sbt.xMain$.$anonfun$run$8(Main.scala:102)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)
	at scala.Console$.withIn(Console.scala:230)
	at sbt.internal.util.Terminal$.withIn(Terminal.scala:560)
	at sbt.internal.util.Terminal$.$anonfun$withStreams$1(Terminal.scala:350)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)
	at scala.Console$.withOut(Console.scala:167)
	at sbt.internal.util.Terminal$.$anonfun$withOut$2(Terminal.scala:550)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)
	at scala.Console$.withErr(Console.scala:196)
	at sbt.internal.util.Terminal$.withOut(Terminal.scala:550)
	at sbt.internal.util.Terminal$.withStreams(Terminal.scala:350)
	at sbt.xMain$.run(Main.scala:86)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sbt.internal.XMainConfiguration.run(XMainConfiguration.scala:83)
	at sbt.xMain.run(Main.scala:46)
	at xsbt.boot.Launch$.$anonfun$run$1(Launch.scala:149)
	at xsbt.boot.Launch$.withContextLoader(Launch.scala:176)
	at xsbt.boot.Launch$.run(Launch.scala:149)
	at xsbt.boot.Launch$.$anonfun$apply$1(Launch.scala:44)
	at xsbt.boot.Launch$.launch(Launch.scala:159)
	at xsbt.boot.Launch$.apply(Launch.scala:44)
	at xsbt.boot.Launch$.apply(Launch.scala:21)
	at xsbt.boot.Boot$.runImpl(Boot.scala:78)
	at xsbt.boot.Boot$.run(Boot.scala:73)
	at xsbt.boot.Boot$.main(Boot.scala:21)
	at xsbt.boot.Boot.main(Boot.scala)
[error] Switch failed: no subprojects list "3.1.2" (or compatible version) in crossScalaVersions setting.
[error] If you want to force it regardless, call ++ 3.1.2!
[error] Use 'last' for the full log.
```

something very wrong:
```
sbt:root> crossScalaVersions
[info] scalatestDiagrams / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalatestWordSpec / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalatestFreeSpec / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalatestFunSuite / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalatestFunSpec / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalatestFeatureSpec / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalatestMatchersCore / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalatestPropSpec / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalatestRefSpec / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalatestShouldMatchers / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalatestMustMatchers / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalatestFlatSpec / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalatestCore / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalatest / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalactic-test / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalacticMacro / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalactic / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] common-test / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalatestFlatSpecTest / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalatestWordSpecTest / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalatestFunSpecTest / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalatestDiagramsTest / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalatestPropSpecTest / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalatestFunSuiteTest / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalatestFeatureSpecTest / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalatestFreeSpecTest / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] scalatest-test / crossScalaVersions
[info] 	List(2.13.6, 2.12.15, 2.11.12, 2.10.7)
[info] crossScalaVersions
[info] 	List(2.12.12)
sbt:root> 
```

**Comment by catap** [comment]
I'm using `06c1d67a1582c4f9d19bb9fdd899a1d765ceccb1` as local root and run it under clean git repository.

**Comment by catap** [comment]
@cheeseng on my machine I have:
```
[error] Found 8 missing definitions while linking
[error] Not found Top(org.scalatest.SuitesL20org.scalatest.SuitesL30scala.collection.immutable.Seq$scalanative$ReflectivelyInstantiate$)
[error] Not found Member(Top(org.scalatest.SuitesL20org.scalatest.SuitesL30scala.collection.immutable.Seq$scalanative$ReflectivelyInstantiate$),RE)
[error] 	at /home/cheeseng/git/scalatest/dotty/core.native/target/scala-3/src_managed/main/org/scalatest/Suites.scala:57
[error] Not found Top(org.scalatest.funspec.AnyFunSpecL32org.scalatest.funspec.AnyFunSpec$scalanative$ReflectivelyInstantiate$)
[error] Not found Member(Top(org.scalatest.funspec.AnyFunSpecL32org.scalatest.funspec.AnyFunSpec$scalanative$ReflectivelyInstantiate$),RE)
[error] 	at /home/cheeseng/git/scalatest/dotty/funspec.native/target/scala-3/src_managed/main/org/scalatest/funspec/AnyFunSpec.scala:1631
[error] Not found Top(org.scalatest.funsuite.AnyFunSuiteL34org.scalatest.funsuite.AnyFunSuite$scalanative$ReflectivelyInstantiate$)
[error] Not found Member(Top(org.scalatest.funsuite.AnyFunSuiteL34org.scalatest.funsuite.AnyFunSuite$scalanative$ReflectivelyInstantiate$),RE)
[error] 	at /home/cheeseng/git/scalatest/dotty/funsuite.native/target/scala-3/src_managed/main/org/scalatest/funsuite/AnyFunSuite.scala:1563
[error] Not found Top(org.scalatest.tools.FrameworkL29org.scalatest.tools.Framework$scalanative$ReflectivelyInstantiate$)
[error] Not found Member(Top(org.scalatest.tools.FrameworkL29org.scalatest.tools.Framework$scalanative$ReflectivelyInstantiate$),RE)
[error] 	at /home/cheeseng/git/scalatest/dotty/core.native/target/scala-3/src_managed/main/org/scalatest/tools/Framework.scala:24
[error] Undefined definitions found in reachability phase
[error] (scalatestPlusScalaCheckNative / Test / nativeLink) Undefined definitions found in reachability phase
```

=> seems you have published wrong paths. Try to understand why.

**Comment by cheeseng** [comment]
@catap I think I found the cause, for Scala 3 it seems to have this extra package in the compiled classes: 

![Screenshot from 2022-04-21 09-51-35](https://user-images.githubusercontent.com/1322009/164355900-9c7176de-dda9-4af9-a109-4eadd525fbe5.png)

I'll include it in the osgi export package list and try again, I'll drop the staged repo I sent you earlier.

**Comment by cheeseng** [comment]
@catap Fyi seems to fix with the following commit: 

https://github.com/scalatest/scalatest/pull/2121/commits/f4e133c4661bc2b79df3e0d563b4e04b3dafe5da

Had to use wildcard for the osgi export package settings to include the generated package mentioned above, a bit weird that it is not needed for scala 2 build though.

Anyway I'll re-stage 3.2.12 release today, as I think scala-js needs this release also.

Thanks for your help!

**Comment by catap** [comment]
@cheeseng I glade that you figured it out. I just re-published it from scratch on my machine to start digging :)

**Comment by lolgab** [comment]
@cheeseng Can this issue be closed now? It seems to be solved by scalatest `3.2.12`

**Comment by cheeseng** [comment]
Build included in 3.2.12.  Thanks.
